pkg/mcp/server/monitoring.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2018 Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package server
+
+import (
+	"go.opencensus.io/stats"
+	"go.opencensus.io/tag"
+	"go.opencensus.io/stats/view"
+)
+
+const (
+	typeURL   = "typeURL"
+	errorCode = "code"
+	errorStr  = "error"
+	connectionID = "connectionID"
+)
+
+var (
+	TypeURLTag      tag.Key
+	ErrorCodeTag    tag.Key
+	ErrorTag        tag.Key
+	ConnectionIDTag tag.Key
+
+	// buckets are powers of 4
+	byteBuckets = []float64{1, 4, 16, 64, 256, 1024, 4096, 16384, 65536, 262144, 1048576, 4194304, 16777216, 67108864, 268435456, 1073741824}
+
+	ClientCount = stats.Int64(
+		"mcp/server/client_count",
+		"The number of clients currently connected.",
+		stats.UnitDimensionless)
+
+	RequestSizesBytes = stats.Int64(
+		"mcp/server/message_sizes_bytes",
```

Comment:
Are these metrics qualified by the source component identity? `istio.io/pkg/mcp/server` is a common package that may be used by Galley _and_ other out-of-processor pilot adapters. It would be good to distinguish the source type (galley vs. oop adapter) and source instance (galley replicas).

Reply:
@douglas-reid I'm not sure how this might be done. Any ideas?

Reply:
@tangiel one way: just prepend `galley/` to the current names you have.

Reply:
I think I misunderstand something here. @ayj is the goal here essentially to be able to prepend any string including `galley/`?

Reply:
The MCP server and client are common packages potentially used by more than one component / repo, e.g. Galley, Pilot, Mixer, CloudFoundry. 

We could prefix the names (e.g. `galley/mcp/server/message_sizes_bytes`). I think it would need to be programmatically set (vs. fixed global variable). We might need to introduce a `StatsContext` struct that carries the the various opencensus stats measures for each particular MCP client/server type, e.g.

```golang
type StatsContext struct {
    requestAcks stats.Int64Measure
    requestNacks stats.Int64Measure
    // etc 
}

func newStatsContext(prefix string) *StatsContext {
    // (1) verify prefix is not reused across stats context? make treat contexts as a singleton per prefix?
    // (2) create instance of StatsContext and corresponding `stats.Int*` measures.
    // (3) view.Register() the new views for this context   
}

// create StatsContext inside of server.New() and client.New() methods.
// These New() functions would need to accept a "prefix" argument which 
// could also be used for logging purposes.
```

Reply:
Agreed that (1) is messy.

re (2), `istio.io/pkg/mcp/server::Server` doesn't have a `Close()` method. We could add one to help with bookkeeping and graceful connection teardown. We'd probably need to refactor the snapshot package as well to consistently handle closing client watches.

Another option (4) is to abstract the metric collection from the MCP server via an interface, e.g.

```golang
type Reporter interface {
    ReportMessageSizeBytes(v int64)
    // etc
}

// in pkg/mcp/server/server.go
func New(watcher Watcher, supportedTypes []string, authChecker authChecker, reporter Reporter) *Server {
    // etc
}
```

For most MCP/Galley tests we can provide a fake reporter that doesn't report to opencensus. Instead, we can collect metrics in maps, etc. This would also allow us better unit test coverage of telemetry without relying on external dependencies.

Reply:
What would be the correct way of parameterizing this for Galley? Would it be correct to augment `patchTable`?

Reply:
Extending `patchTable` sounds reasonable.